File: hggzkportal_app/lib/features/admin_bookings/domain/entities/booking.dart
import 'package:equatable/equatable.dart';
import '../../../../../core/enums/booking_status.dart';

/// 📋 Entity الحجز الأساسي
class Booking extends Equatable {
  final String id;
  final String userId;
  final String unitId;
  final DateTime checkIn;
  final DateTime checkOut;
  final int guestsCount;
  final Money totalPrice;
  final BookingStatus status;
  final DateTime bookedAt;
  final String userName;
  final String unitName;

  // معلومات إضافية
  final String? userEmail;
  final String? userPhone;
  final String? unitImage;
  final String? propertyId;
  final String? propertyName;
  final String? notes;
  final String? specialRequests;
  final String? cancellationReason;
  final DateTime? cancelledAt;
  final DateTime? confirmedAt;
  final DateTime? checkedInAt;
  final DateTime? checkedOutAt;
  final String? bookingSource;
  final bool? isWalkIn;
  final String? paymentStatus;

  const Booking({
    required this.id,
    required this.userId,
    required this.unitId,
    required this.checkIn,
    required this.checkOut,
    required this.guestsCount,
    required this.totalPrice,
    required this.status,
    required this.bookedAt,
    required this.userName,
    required this.unitName,
    this.userEmail,
    this.userPhone,
    this.unitImage,
    this.propertyId,
    this.propertyName,
    this.notes,
    this.specialRequests,
    this.cancellationReason,
    this.cancelledAt,
    this.confirmedAt,
    this.checkedInAt,
    this.checkedOutAt,
    this.bookingSource,
    this.isWalkIn,
    this.paymentStatus,
  });

  /// حساب عدد الليالي
  int get nights {
    return checkOut.difference(checkIn).inDays;
  }

  /// التحقق من إمكانية الإلغاء
  bool get canCancel {
    return status == BookingStatus.pending || status == BookingStatus.confirmed;
  }

  /// التحقق من إمكانية التأكيد
  bool get canConfirm {
    return status == BookingStatus.pending;
  }

  /// التحقق من إمكانية تسجيل الوصول
  bool get canCheckIn {
    return status == BookingStatus.confirmed &&
        DateTime.now().isAfter(checkIn.subtract(const Duration(hours: 2)));
  }

  /// التحقق من إمكانية تسجيل المغادرة
  bool get canCheckOut {
    return status == BookingStatus.checkedIn;
  }

  Booking copyWith({
    String? id,
    String? userId,
    String? unitId,
    DateTime? checkIn,
    DateTime? checkOut,
    int? guestsCount,
    Money? totalPrice,
    BookingStatus? status,
    DateTime? bookedAt,
    String? userName,
    String? unitName,
    String? userEmail,
    String? userPhone,
    String? unitImage,
    String? propertyId,
    String? propertyName,
    String? notes,
    String? specialRequests,
    String? cancellationReason,
    DateTime? cancelledAt,
    DateTime? confirmedAt,
    DateTime? checkedInAt,
    DateTime? checkedOutAt,
    String? bookingSource,
    bool? isWalkIn,
    String? paymentStatus,
  }) {
    return Booking(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      unitId: unitId ?? this.unitId,
      checkIn: checkIn ?? this.checkIn,
      checkOut: checkOut ?? this.checkOut,
      guestsCount: guestsCount ?? this.guestsCount,
      totalPrice: totalPrice ?? this.totalPrice,
      status: status ?? this.status,
      bookedAt: bookedAt ?? this.bookedAt,
      userName: userName ?? this.userName,
      unitName: unitName ?? this.unitName,
      userEmail: userEmail ?? this.userEmail,
      userPhone: userPhone ?? this.userPhone,
      unitImage: unitImage ?? this.unitImage,
      propertyId: propertyId ?? this.propertyId,
      propertyName: propertyName ?? this.propertyName,
      notes: notes ?? this.notes,
      specialRequests: specialRequests ?? this.specialRequests,
      cancellationReason: cancellationReason ?? this.cancellationReason,
      cancelledAt: cancelledAt ?? this.cancelledAt,
      confirmedAt: confirmedAt ?? this.confirmedAt,
      checkedInAt: checkedInAt ?? this.checkedInAt,
      checkedOutAt: checkedOutAt ?? this.checkedOutAt,
      bookingSource: bookingSource ?? this.bookingSource,
      isWalkIn: isWalkIn ?? this.isWalkIn,
      paymentStatus: paymentStatus ?? this.paymentStatus,
    );
  }

  @override
  List<Object?> get props => [
        id,
        userId,
        unitId,
        checkIn,
        checkOut,
        guestsCount,
        totalPrice,
        status,
        bookedAt,
        userName,
        unitName,
        userEmail,
        userPhone,
        unitImage,
        propertyId,
        propertyName,
        notes,
        specialRequests,
        cancellationReason,
        cancelledAt,
        confirmedAt,
        checkedInAt,
        checkedOutAt,
        bookingSource,
        isWalkIn,
        paymentStatus,
      ];
}

/// 💰 Entity للمبلغ المالي
class Money extends Equatable {
  final double amount;
  final String currency;
  final String formattedAmount;

  const Money({
    required this.amount,
    required this.currency,
    required this.formattedAmount,
  });

  Money copyWith({
    double? amount,
    String? currency,
    String? formattedAmount,
  }) {
    return Money(
      amount: amount ?? this.amount,
      currency: currency ?? this.currency,
      formattedAmount: formattedAmount ?? this.formattedAmount,
    );
  }

  @override
  List<Object> get props => [amount, currency, formattedAmount];
}

File: backend/YemenBooking.Core/Entities/Booking.cs
namespace YemenBooking.Core.Entities;

using System;
using YemenBooking.Core.ValueObjects;
using System.Collections.Generic;
using YemenBooking.Core.Enums;
using System.ComponentModel.DataAnnotations;

/// <summary>
/// كيان الحجز
/// Booking entity
/// </summary>
[Display(Name = "كيان الحجز")]
public class Booking : BaseEntity<Guid>
{
    /// <summary>
    /// معرف المستخدم
    /// User identifier
    /// </summary>
    [Display(Name = "معرف المستخدم")]
    public Guid UserId { get; set; }
    
    /// <summary>
    /// معرف الوحدة
    /// Unit identifier
    /// </summary>
    [Display(Name = "معرف الوحدة")]
    public Guid UnitId { get; set; }
    
    /// <summary>
    /// تاريخ الوصول
    /// Check-in date
    /// </summary>
    [Display(Name = "تاريخ الوصول")]
    public DateTime CheckIn { get; set; }
    
    /// <summary>
    /// تاريخ المغادرة
    /// Check-out date
    /// </summary>
    [Display(Name = "تاريخ المغادرة")]
    public DateTime CheckOut { get; set; }
    
    /// <summary>
    /// عدد الضيوف
    /// Number of guests
    /// </summary>
    [Display(Name = "عدد الضيوف")]
    public int GuestsCount { get; set; }
    
    /// <summary>
    /// السعر الإجمالي للحجز
    /// Total price of the booking
    /// </summary>
    [Display(Name = "السعر الإجمالي للحجز")]
    public Money TotalPrice { get; set; }
    
    /// <summary>
    /// حالة الحجز (مؤكّد، معلّق، ملغى)
    /// Booking status (Confirmed, Pending, Cancelled)
    /// </summary>
    [Display(Name = "حالة الحجز")]
    public BookingStatus Status { get; set; }
    
    /// <summary>
    /// تاريخ الحجز
    /// Booking date
    /// </summary>
    [Display(Name = "تاريخ الحجز")]
    public DateTime BookedAt { get; set; }

    /// <summary>
    /// مصدر الحجز (ويب، موبايل، حجز مباشر)
    /// Booking source (WebApp, MobileApp, WalkIn)
    /// </summary>
    [Display(Name = "مصدر الحجز")]
    public string? BookingSource { get; set; }

    /// <summary>
    /// سبب الإلغاء
    /// Cancellation reason
    /// </summary>
    [Display(Name = "سبب الإلغاء")]
    public string? CancellationReason { get; set; }

    /// <summary>
    /// هل الحجز مباشر (Walk-in)
    /// Is walk-in booking
    /// </summary>
    [Display(Name = "هل الحجز مباشر")]
    public bool IsWalkIn { get; set; } = false;
    
    /// <summary>
    /// المستخدم المرتبط بالحجز
    /// User associated with the booking
    /// </summary>
    [Display(Name = "المستخدم المرتبط بالحجز")]
    public virtual User User { get; set; }
    
    /// <summary>
    /// الوحدة المرتبطة بالحجز
    /// Unit associated with the booking
    /// </summary>
    [Display(Name = "الوحدة المرتبطة بالحجز")]
    public virtual Unit Unit { get; set; }
    
    /// <summary>
    /// المدفوعات المرتبطة بالحجز
    /// Payments associated with the booking
    /// </summary>
    [Display(Name = "المدفوعات المرتبطة بالحجز")]
    public virtual ICollection<Payment> Payments { get; set; } = new List<Payment>();
    
    /// <summary>
    /// الخدمات المرتبطة بالحجز
    /// Services associated with the booking
    /// </summary>
    [Display(Name = "الخدمات المرتبطة بالحجز")]
    public virtual ICollection<BookingService> BookingServices { get; set; } = new List<BookingService>();
    
    /// <summary>
    /// المراجعات المرتبطة بالحجز
    /// Reviews associated with the booking
    /// </summary>
    [Display(Name = "المراجعات المرتبطة بالحجز")]
    public virtual ICollection<Review> Reviews { get; set; } = new List<Review>();

    /// <summary>
    /// الجداول اليومية المتعلقة بالحجز
    /// Daily schedules related to this booking
    /// </summary>
    public virtual ICollection<DailyUnitSchedule> DailySchedules { get; set; } = new List<DailyUnitSchedule>();

    /// <summary>
    /// مبلغ عمولة المنصة
    /// Platform commission amount
    /// </summary>
    [Display(Name = "مبلغ عمولة المنصة")]
    public decimal PlatformCommissionAmount { get; set; }

    /// <summary>
    /// تاريخ تسجيل الوصول الفعلي
    /// Actual check-in date
    /// </summary>
    [Display(Name = "تاريخ تسجيل الوصول الفعلي")]
    public DateTime? ActualCheckInDate { get; set; }

    /// <summary>
    /// تاريخ المغادرة الفعلي
    /// Actual check-out date
    /// </summary>
    [Display(Name = "تاريخ المغادرة الفعلي")]
    public DateTime? ActualCheckOutDate { get; set; }

    /// <summary>
    /// المبلغ النهائي المدفوع
    /// Final amount
    /// </summary>
    [Display(Name = "المبلغ النهائي المدفوع")]
    public decimal FinalAmount { get; set; }

    /// <summary>
    /// تقييم العميل
    /// Customer rating
    /// </summary>
    [Display(Name = "تقييم العميل")]
    public decimal? CustomerRating { get; set; }

    /// <summary>
    /// ملاحظات إكمال الحجز
    /// Completion notes
    /// </summary>
    [Display(Name = "ملاحظات إكمال الحجز")]
    public string? CompletionNotes { get; set; }

    /// <summary>
    /// لقطة سياسات العقار وقت إنشاء الحجز (JSON)
    /// Property policy snapshot at booking time (JSON)
    /// </summary>
    [Display(Name = "لقطة سياسات العقار وقت إنشاء الحجز (JSON)")]
    public string? PolicySnapshot { get; set; }

    /// <summary>
    /// تاريخ حفظ لقطة السياسات
    /// When the policy snapshot was captured
    /// </summary>
    [Display(Name = "تاريخ حفظ لقطة السياسات")]
    public DateTime? PolicySnapshotAt { get; set; }
}